main/src/ca/uwaterloo/flix/language/phase/Typer.scala

```diff
@@ -1247,7 +1247,12 @@ object Typer extends Phase[ResolvedAst.Program, TypedAst.Root] {
           constraintsType2 <- unifyTypM(constraintsType, Type.SchemaExtend(sym, freshPredicateTypeVar, freshRestTypeVar), loc)
           tupleType = sym match {
             case rel: Symbol.RelSym =>
-              Type.mkTuple(program.relations(rel).attr.map(a => a.tpe))
+              val tpes = program.relations(rel).attr.map(a => a.tpe)
+              tpes match {
+                case Nil => mkUnitType() // If there's no field in the relation, use Unit
+                case tpe :: Nil => tpe // If there's a single field, use its type
+                case _ => Type.mkTuple(tpes) // If there are more fields, use a tuple
+              }
```

Comment:
We are moving away from having global declarations of relations/lattices, so we need some other mechanism to determine the length. Any ideas?

Reply:
The type of the predicate should be a tuple, i.e. what is called `freshPredicateTypeVar` here. (which seem unused (?)) 

Here the tuple is constructed:

https://github.com/flix/flix/blob/master/main/src/ca/uwaterloo/flix/language/phase/Resolver.scala#L330

so I think *here* you want to special case like you have done above. Does that make sense?

Reply:
It does make sense but I feel like I'm still missing some understanding here. `freshPredicateTypeVar` is used to unify with `constraintsType` a few lines higher in the typer:
```
// constraints should have the form {pred.sym : freshPredicateTypeVar | freshRestTypeVar}
constraintsType2 <- unifyTypM(constraintsType, Type.SchemaExtend(sym, freshPredicateTypeVar, freshRestTypeVar), loc)
```

Is there a way to retrieve what is has been unified to after the facts? (Just accessing `freshPredicateTypeVar` gives a type variable, while I need to access the tuple/relation type)

Reply:
Let us talk about it on Skype.

Reply:
I made progress on that, but haven't run all the tests before pushing. One test is broken (boxed exception), I'll come back to it.